shared/include/userver/utils/fixed_array.hpp

```diff
@@ -63,18 +77,29 @@ class FixedArray final {
   const T* cend() const noexcept { return data() + size_; }
 
  private:
+  T* NonEmptyData() noexcept {
+    UASSERT(size_);
+    return data();
+  }
+
+  const T* NonEmptyData() const noexcept {
+    UASSERT(size_);
+    return data();
+  }
+
   T* storage_;
   std::size_t size_;
 };
 
+template <class Range>
+FixedArray(FromRangeTag tag, Range&& range)
+    ->FixedArray<std::decay_t<decltype(*range.begin())>>;
+
 template <class T>
 template <class... Args>
 FixedArray<T>::FixedArray(std::size_t size, Args&&... args)
     : storage_(std::allocator<T>{}.allocate(size)), size_(size) {
-  UASSERT(size > 0);
-  if (!size) {
-    return;
-  }
+  if (size == 0) return;
```

Comment:
No need to call allocate if size is 0. It is better to initialize `storage_` to nullptr, do the check and then allocate